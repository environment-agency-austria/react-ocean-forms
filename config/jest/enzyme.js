import Enzyme from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';
import 'babel-polyfill';

// Currently there's no official enzyme support
// for reacts context api. FIXME
// https://github.com/airbnb/enzyme/issues/1553
import Adapter from './enzymeAdapter';

Enzyme.configure({ adapter: new Adapter() });
