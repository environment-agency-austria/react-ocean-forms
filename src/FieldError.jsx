/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';

import { toArray } from './utils';
import { errorsShape } from './shapes';


/**
 * Component for displaying bootstrap
 * form feedbacks if there are any errors
 */
function FieldError(props) {
  const {
    id,
    invalid,
    error,
    stringFormatter,
  } = props;

  // If the field isn't invalid do nothing
  if (invalid !== true || error === null) return null;

  // Error could be either an string or an array of strings
  const errorArray = toArray(error);
  const errors = errorArray.map((item) => {
    const errorString = stringFormatter(item.message_id, item.params);
    return <span key={`${id}_${item.message_id}`}>{errorString}</span>;
  });

  return errors;
}

FieldError.defaultProps = {
  error: null,
};

FieldError.propTypes = {
  id: PropTypes.string.isRequired,
  invalid: PropTypes.bool.isRequired,
  error: errorsShape,
  stringFormatter: PropTypes.func.isRequired,
};

export default FieldError;
