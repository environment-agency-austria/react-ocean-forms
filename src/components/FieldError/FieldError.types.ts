import { TSTringFormatter } from '../../utils';
import { TFieldErrors } from '../../validators';

/**
 * Props for the FieldError component
 */
export interface IFieldErrorProps {
  /**
   * Html id
   */
  id: string;
  /**
   * True if the field is invalid
   */
  invalid: boolean;
  /**
   * Field errors
   */
  error: TFieldErrors;
  /**
   * stringFormatter method
   */
  stringFormatter: TSTringFormatter;
}
