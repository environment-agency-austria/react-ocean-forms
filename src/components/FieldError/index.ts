export * from './FieldError';
export * from './FieldError.types';
