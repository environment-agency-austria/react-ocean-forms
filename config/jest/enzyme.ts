import enzyme from 'enzyme';
import enzymeAdapterReact16 from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new enzymeAdapterReact16() });
