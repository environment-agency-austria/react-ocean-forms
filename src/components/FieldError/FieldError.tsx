/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { useFormContext } from '../../hooks';
import { IFieldErrorProps } from './FieldError.types';

/**
 * Component for displaying bootstrap
 * form feedbacks if there are any errors
 */
export const FieldError: React.FC<IFieldErrorProps> = (props) => {
  const {
    id,
    invalid,
    error,
  } = props;

  const { stringFormatter } = useFormContext();

  // If the field isn't invalid do nothing
  if (invalid !== true || error === null) { return null; }

  // Error could be either an string or an array of strings
  const errorArray = Array.isArray(error) ? error : [error];

  return (
    <>
      {errorArray.map((item) => {
        const errorString = stringFormatter(item.message_id, item.params);

        return <span key={`${id}_${item.message_id}`}>{errorString}</span>;
      })}
    </>
  );
};
FieldError.displayName = 'FieldError';
