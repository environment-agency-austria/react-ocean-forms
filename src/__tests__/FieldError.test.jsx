import React from 'react';
import { shallow } from 'enzyme';

import FieldError from '../FieldError';

describe('<FieldError />', () => {
  const mockStringFormatter = jest.fn().mockReturnValue('string');
  const setup = props => shallow((
    <FieldError
      id="unitError"
      stringFormatter={mockStringFormatter}
      {...props}
    />
  ));

  it('should do nothing if invalid is false', () => {
    const wrapper = setup({ invalid: false });
    expect(wrapper).toMatchSnapshot();
  });

  const errorId = 'foo';
  const errorParams = { foo: 'bar' };

  it('should render an error', () => {
    const wrapper = setup({
      invalid: true,
      error: {
        message_id: errorId,
        params: errorParams,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should correctly use the stringFormatter', () => {
    expect(mockStringFormatter).toHaveBeenLastCalledWith(
      errorId,
      errorParams,
    );
  });
});
