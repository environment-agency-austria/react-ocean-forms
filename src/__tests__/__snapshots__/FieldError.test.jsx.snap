// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldError /> should do nothing if invalid is false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldError
    error={null}
    id="unitError"
    invalid={false}
    stringFormatter={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<FieldError /> should render an error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldError
    error={
      Object {
        "message_id": "foo",
        "params": Object {
          "foo": "bar",
        },
      }
    }
    id="unitError"
    invalid={true}
    stringFormatter={
      [MockFunction] {
        "calls": Array [
          Array [
            "foo",
            Object {
              "foo": "bar",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": "string",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "unitError_foo",
    "nodeType": "host",
    "props": Object {
      "children": "string",
    },
    "ref": null,
    "rendered": "string",
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "unitError_foo",
      "nodeType": "host",
      "props": Object {
        "children": "string",
      },
      "ref": null,
      "rendered": "string",
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
